operator/pkg/bgpv2/bgpp.go

```diff
@@ -0,0 +1,503 @@
+		},
+		Spec: cilium_api_v2alpha1.CiliumBGPNodeSpec{
+			BGPInstances: nodeBGPInstancesFromBGPP(bgpp),
+		},
+	}
+
+	switch {
+	case exists && prev.Spec.DeepEqual(&nodeConfig.Spec):
+		return nil
+	case exists:
+		prev.Spec = nodeConfig.Spec
+		_, err = b.nodeConfigClient.Update(ctx, prev, meta_v1.UpdateOptions{})
+	default:
+		_, err = b.nodeConfigClient.Create(ctx, nodeConfig, meta_v1.CreateOptions{})
+		if err != nil && k8s_errors.IsAlreadyExists(err) {
+			// local store is not yet updated, but API server has the resource. Get resource from API server
+			// to compare spec.
+			prev, err = b.nodeConfigClient.Get(ctx, nodeConfig.Name, meta_v1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			// if prev already exist and spec is different, update it
+			if !prev.Spec.DeepEqual(&nodeConfig.Spec) {
+				prev.Spec = nodeConfig.Spec
+				_, err = b.nodeConfigClient.Update(ctx, prev, meta_v1.UpdateOptions{})
+			} else {
+				return nil
+			}
+		}
+	}
+
+	b.logger.WithField("node config", nodeRef).Debug("Upserting BGP Node Config")
+
+	return err
+}
+
+// peerConfigRefFromBGPP returns a PeerConfigReference from the given policyName, localASN and peerAddr.
+func peerConfigRefFromBGPP(policyName string, localASN int64, peerAddr string) *cilium_api_v2alpha1.PeerConfigReference {
+	return &cilium_api_v2alpha1.PeerConfigReference{
+		Group: cilium_api_v2alpha1.CustomResourceDefinitionGroup,
+		Kind:  cilium_api_v2alpha1.BGPPCKindDefinition,
+		Name:  peerKeyFromBGPP(policyName, localASN, peerAddr),
+	}
+}
+
+// peerKeyFromBGPP returns a key for the given bgp policy, localASN and peerAddr.
+// Note returned string is used as name in various kubernetes resources, so naming convention should be followed.
+// Naming should consist of lower case alphanumeric characters, -, and .
+func peerKeyFromBGPP(policyName string, localASN int64, peerAddr string) string {
+	peerAddr = strings.ReplaceAll(peerAddr, "/", "-")
+	peerAddr = strings.ReplaceAll(peerAddr, ":", ".")
+	key := fmt.Sprintf("%s-%s", instanceKeyFromBGPP(policyName, localASN), peerAddr)
+	return strings.ToLower(key)
+}
+
+func instanceKeyFromBGPP(policyName string, localASN int64) string {
+	return fmt.Sprintf("%s-%d", policyName, localASN)
+}
+
+// nodeBGPInstancesFromBGPP returns a list of CiliumBGPNodeInstance from the given CiliumBGPPeeringPolicy.
+func nodeBGPInstancesFromBGPP(bgpp *cilium_api_v2alpha1.CiliumBGPPeeringPolicy) []cilium_api_v2alpha1.CiliumBGPNodeInstance {
+	var res []cilium_api_v2alpha1.CiliumBGPNodeInstance
+
+	for _, vr := range bgpp.Spec.VirtualRouters {
+		vr := vr
+		var instance cilium_api_v2alpha1.CiliumBGPNodeInstance
+		instance.Name = instanceKeyFromBGPP(bgpp.Name, vr.LocalASN)
+		instance.LocalASN = &vr.LocalASN
+
+		for _, neigh := range vr.Neighbors {
+			neigh := neigh
+
+			// BGP Peering Policy peer address is /32 or /128 CIDR. We transform it to IP address.
+			// On failure, skip the neighbor.
+			peerAddr, err := prefixToAddr(neigh.PeerAddress)
```

Comment:
Should `prefixToAddr()` check that the mask length is 32 or 128?

Reply:
Thinking more about it, such validation is going to be tricky with regex. I'll create a follow up PR to add this check here and unit tests. ( just to keep scope of this PR limited as it's getting quite large ).